sale_order_merge/models/sale_order_merge.py

```diff
@@ -0,0 +1,190 @@
+        'sale.order', 'rel_sale_to_merge', 'sale_id', 'to_merge_id',
+        'Orders to merge')
+
+    @api.multi
+    def merge_order_lines(self):
+        self.sale_order.write({
+            'order_line': [
+                (4, line.id)
+                for line in self.to_merge.mapped('order_line')
+            ]})
+
+    @api.multi
+    def merge_invoices(self):
+        """ Merge all draft invoices. For prepaid orders, the payment
+        of the original invoice is leading to start the procurement, but
+        there may still be other confirmed invoices. """
+        target = self.env['account.invoice']
+        other_inv = self.env['account.invoice']
+        keep_inv = self.env['account.invoice']
+        for invoice in (
+                self.sale_order.invoice_ids +
+                self.to_merge.mapped('invoice_ids')):
+            if invoice.state == 'draft' and not invoice.internal_number:
+                if target:
+                    other_inv += invoice
+                else:
+                    target = invoice
+            else:
+                keep_inv += invoice
+        if target:
+            other_inv.mapped('invoice_line').write({'invoice_id': target.id})
+            other_inv.mapped('tax_line').write({'invoice_id': target.id})
+            other_inv.unlink()
+            target.button_compute(set_total=True)
+
+        for inv in target + keep_inv:
+            self.sale_order.write({'invoice_ids': [(4, inv.id)]})
+        self.to_merge.write({'invoice_ids': [(6, 0, [])]})
+
+    @api.multi
+    def merge_pickings(self):
+        """ Assign all pickings to the target sale order and merge any
+        pending pickings """
+        orders = self.sale_order + self.to_merge
+        group = self.env['procurement.group']
+        if self.sale_order.procurement_group_id:
+            group = self.sale_order.procurement_group_id
+        else:
+            for order in self.to_merge:
+                if order.procurement_group_id:
+                    group = order.procurement_group_id
+                    break
+            else:
+                return  # no group, no pickings
+            self.sale_order.write({'procurement_group_id': group.id})
+        other_groups = orders.mapped('procurement_group_id')
+        self.env['stock.picking'].search(
+            [('group_id', 'in', other_groups.ids)]).write(
+                {'group_id': group.id})
+        self.env['stock.move'].search(
+            [('group_id', 'in', other_groups.ids)]).write(
+                {'group_id': group.id})
+        self.env['procurement.order'].search(
+            [('group_id', 'in', other_groups.ids)]).write(
+                {'group_id': group.id})
+        pick_map = {}
+        for picking in self.sale_order.picking_ids:
+            if (picking.state not in ('done', 'cancel') and
+                    picking.location_dest_id.usage == 'customer'):
+                key = (picking.picking_type_id, picking.location_id,
```

Comment:
Looking towards extensibility, please build the key in a separate method. We work, for instance, on another model that various pickings for a sales order, based on the requested date maintained at sales order line level. In this case, when we merge orders we'd want to respect the scheduled date.